locked build/hot reload

padraicbc opened the pull request:
* I ended up taking a different approach where we lock on build. For hot reloading the websocket tells us when onload fires so we know it is ok to build again. 
    * If there is no lock then the refresh can happen in the midst of another build  which can break the app. If not using hotreload then we just unlock after Build is complete. 
* I think locking the build is the way to go regardless as you have race conditions with multiple read/write access to the same files. 
The logic may have to be tweaked when we add  concurrency to the build but it should pretty reliable to use as is for now.


* To use hot reloading just pass the flag to serve:
        `plenti serve -H/--hot-reload`

jimafisk commented:
Awesome, super excited for this! I'm slammed with client work at the moment, but hope to take a look soon. Thanks for grabbing this and running with it!

padraicbc commented:
> Awesome, super excited for this! I'm slammed with client work at the moment, but hope to take a look soon. Thanks for grabbing this and running with it!

No worries.  I still have to add broadcasting so it works when multiple tabs/browsers are open where all connected sockets get pinged to reload. Should get that integrated in next day or 2 then hopefully should be good to go.

padraicbc commented:
The latest handles multiple tabs/browsers. It is not trivial when you have multiple connections as you have to wait for builds complete, then signal reload to all and finally wait for them all to signal that they are loaded. If not you will end up in a situation where one can get caught in the middle of another build and break.  

Having spent time on this it really highlights the need to separate the full build vs develpement. The build `log.Fatal's` which is correct as you don't want anything broken in your production build. 
That  experience is less desirable when developing. The full build can take quite a while also even with a few files so putting together a means to separate the full production build and local dev is definitely the key to a nice developer experience.

There are a few changes that I think would greatly improve the reloading.

* Create a separate build process as mentioned above.
*  Put all the staged builds/changes on a channel and drain that channel if they come in quick succession after beginning the initial although less of an issue if we implement faster partial builds.
* Have an error channel that accessible throughout the build that can signal so we don't get locked in a build on error. 
* Handle those errors gracefully and inject a socket even on broken builds to allow a quick fix/save and reload.

padraicbc commented:
There are quite a few changes in the latest. 

- When using serve build errors won't `log.Fatal`
- There is more output added to error messages like the effected filename and line number ithe error returned from. It was more for my own debugging but may be useful.
- Now when we have a build we use atomic seen in  [common](https://github.com/padraicbc/plenti/blob/4c8f603f30358ef09a734b1fe5651f725cbefd38/common/errs.go#L21) to allow a check to determine if there is an ongoing build. If there is we just ignore the change. Initially I had locked and queued but it was a mess when you had a few quick saves. Using `CompareAndSwapUint` is preferable to a mutex as you can actually check if it is locked. With a mutex you will get a panic tryin to lock an already locked mutex and vice-versa.

jimafisk commented:
@padraicbc I just checked this out locally and it's looking awesome! Thanks for leaving the detailed notes, I'm hitting the sack for the night but I'm excited to dig into this deeper tomorrow!

jimafisk commented on common/errs.go:
```diff
@@ -1,10 +1,84 @@
+}
+
+// Caller shows the line in was from.
+func Caller() string {
```
This is very helpful!

jimafisk commented on cmd/build/data_source.go:
```diff
@@ -1,17 +1,74 @@
+// It gets set using server flags.
+var Doreload bool
+
+// inject wherever/however, this is just to get it working.
```
Yeah we'll possibly want to move this somewhere, it's fine here for now

jimafisk commented on cmd/serve.go:
```diff
@@ -119,6 +131,8 @@ func init() {
 	serveCmd.Flags().BoolVarP(&VerboseFlag, "verbose", "v", false, "show log messages")
 	serveCmd.Flags().BoolVarP(&BenchmarkFlag, "benchmark", "b", false, "display build time statistics")
 	serveCmd.Flags().BoolVarP(&SSLFlag, "ssl", "s", false, "ssl/tls encryption to serve localhost over https")
+	serveCmd.Flags().BoolVarP(&build.Doreload, "hot-reload", "H", false, "Enable hot reload")
```
I've been using the terms "hot reload" and "live reload" interchangeable, zaydek had a [good point](https://github.com/plentico/plenti/issues/99#issuecomment-773995500) though:

> HMR as in when state is persisted between refreshes, live reload as in hard reloads

Do you think we should change this to `--live-reload` / `-L` to prevent confusion with HMR?

jimafisk commented:
How do you think we're looking on this @padraicbc? It seems to work great for me when testing locally, are there still some things you're working on before you want to merge it in? I'd def be down to chat if you're interested, feel free to DM me on twitter (https://twitter.com/jimafisk) or send an email to plentico@jantcu.com if you'd like to connect. Thanks!

padraicbc commented:
> How do you think we're looking on this @padraicbc? It seems to work great for me when testing locally, are there still some things you're working on before you want to merge it in? I'd def be down to chat if you're interested, feel free to DM me on twitter (https://twitter.com/jimafisk) or send an email to [plentico@jantcu.com](mailto:plentico@jantcu.com) if you'd like to connect. Thanks!

@jimafisk I tidied up the error handling and added a bit more output to the error messages. I squashed it into one commit there so all the changes are visible. 
There are some edge cases with multiple devices/tabs/browsers being open where closing one and reloading before a ping check for life comes back or times out but beyond that it should be good. I have used it concurrently with a few browsers/tabs plus my android devcies and it worked well. 
I will be tweaking the logic of the communicaton as i am not in love with the pinging but that won't effect the end user.

jimafisk commented:
> I will be tweaking the logic of the communicaton as i am not in love with the pinging but that won't effect the end user.

Ok sounds great, just let me know when you want me to merge. It looks great!

padraicbc commented:
You can merge away once you're happy with the latest.  The tweaks will be on the todo list for a while!
